Use interface and classes to support ortholog groups

jmedinaebi opened the pull request:
A few things in this PR:
- Renaming interactor as component everywhere
- Remove references to subcomplexes where possible and just mention subcomponents
- Use enums for sorting and grouping types
- Rename `EnrichedComplex` as `NavigatorComplex`
- Rename `EnrichedInteractor` as `INavigatorComponent`
- Create 2 classes that implement `INavigatorComponent`, one for simple components and one for ortholog groups
- Move logic to create `INavigatorComponent` to before we do the clustering, so we do the clustering after the ortholog groups are created

susiehgt committed:
Starting to work on orthology implementation

susiehgt committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts

susiehgt committed:
Merge remote-tracking branch 'origin/develop' into develop

susiehgt committed:
Merge remote-tracking branch 'origin/develop' into develop

susiehgt committed:
Merge remote-tracking branch 'origin/develop' into develop

susiehgt committed:
Merge remote-tracking branch 'origin/develop' into develop

jmedinaebi committed:
Use interface and classes to support ortholog groups

jmedinaebi committed:
Fix sorting of complexes when using ortholog groups

jmedinaebi committed:
Revert "Fix sorting of complexes when using ortholog groups"

This reverts commit 3da2d1cca549b8018f10ca6141f51c3c1dfefb98.

jmedinaebi committed:
Fix clustering of ortholog groups

jmedinaebi committed:
Make orthologs more visible

jmedinaebi committed:
Only create ortholog groups when there is more than 1 protein

susiehgt committed:
Making predicted complexes more contrasted

jmedinaebi committed:
Fix small issue with lines on expanded components

susiehgt committed:
Change of the orthologs groups display;
Ortholog grouping button available when orthologs groups are present;

jmedinaebi commented on src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.ts:
```diff
@@ -1,24 +1,41 @@
   idDisplay = model<boolean>();
+  complexSearch = input<ComplexSearchResult>();
+
+  orthologButtonAvailable = computed(() => this.checkIfPantherGroups());
```
Try this instead.

1. Delete method `checkIfPantherGroups`

2. Update `orthologButtonAvailable` to be
   ```
   orthologButtonAvailable = input<boolean>();
   ```

3. In `complex-navigator.component.ts` create a new boolean variable to store if there are ortholog groups
   ```
   orthologGroupsAvailable = true;
   ```

4. And pass it to the complex-navigator-buttons component in `complex-navigator-component.html`
   ```
    <cp-complex-navigator-buttons
      [(componentsSorting)]="componentsSorting"
      [(componentsGrouping)]="componentsGrouping"
      [(organismIconDisplay)]="organismIconDisplay"
      [(interactorTypeDisplay)]="interactorTypeDisplay"
      [(idDisplay)]="idDisplay"
      [orthologButtonAvailable]="orthologGroupsAvailable"
      [complexSearch]="complexSearch()">
    </cp-complex-navigator-buttons>
   ```

5. In `table-structure.component.ts` create a model to set the ortholog groups value
   ```
   orthologGroupsAvailable = model<boolean>();
   ```

6. And in that same component, set the value in the `createOrthologGroups` method if we create any group
   ```
    ...
    let anyOrthologGroupCreated = false; // new line
    const newNavigatorComponents: INavigatorComponent[] = [];
    for (const [_, proteins] of Object.entries(groupedByOrthology)) {
      if (proteins.length > 1) {
        // Multiple proteins in the ortholog group
        const group = proteins[0].orthologsGroup;
        const orthologsGroup = new NavigatorOrthologGroup(group, proteins);
        newNavigatorComponents.push(orthologsGroup);
        anyOrthologGroupCreated = true; // new line
      } else if (proteins.length === 1) {
        // Single proteins in the ortholog group, there's no need to create the group, we just use the protein component
        newNavigatorComponents.push(proteins[0]);
      }
    }
    this.orthologGroupsAvailable.set(anyOrthologGroupCreated); // new line
    newNavigatorComponents.push(...interactorsWithoutGroup);
    return newNavigatorComponents;
   ```

7. Finally, in `complex-navigator.component.html` pass the `orthologGroupsAvailable` variable to the model so it can be set
   ```
   <cp-table-structure [complexSearch]="complexSearch()"
                      [interactors]="interactors()"
                      [componentsSorting]="componentsSorting"
                      [componentsGrouping]="componentsGrouping"
                      [organismIconDisplay]="organismIconDisplay"
                      [interactorTypeDisplay]="interactorTypeDisplay"
                      [idDisplay]="idDisplay"
                      [canAddComplexesToBasket]="canAddComplexesToBasket()"
                      [canRemoveComplexesFromBasket]="canRemoveComplexesFromBasket()"
                      [(orthologGroupsAvailable)]="orthologGroupsAvailable"
                      (onComplexRemovedFromBasket)="onComplexRemovedFromBasket.emit($event)">
   </cp-table-structure>
   ```

I haven't tried this but it should work. `TableStructureComponent` sets if there are ortholog groups or not, it passes the value through the model to `ComplexNavigatorComponent` which passes it to `ComplexNavigatorButtonsComponent` to enable or disable the button.

It's like what we do for the sorting or grouping, but the other way around.

susiehgt committed:
Buggy button

jmedinaebi committed:
Move calculation of components one layer up

susiehgt committed:
Fixing Orthology group display inline and compact;

susiehgt committed:
Fixing Orthology group display inline and compact;

jmedinaebi committed:
Update navigatorComponents when research

jmedinaebi committed:
Fix group by orthology

susiehgt committed:
Basket bug fixing;
Small optimisations of the code;

jmedinaebi commented on package.json:
```diff
@@ -37,6 +37,7 @@
     "expose-loader": "^5.0.0",
     "jquery": "^3.7.1",
     "litemol": "https://github.com/dsehnal/LiteMol/tarball/master",
+    "marked": "^12.0.2",
```
What's this dependency for?

susiehgt replied:
My code would not compile if I did not add it, the console was asking me to

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-ortholog-group.model.ts:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  get isOrthologGroup(): boolean {
+    return this._isOrthologGroup;
```
We can just do `return true;`

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-simple-component.model.ts:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  get isOrthologGroup(): boolean {
+    return this._isOrthologGroup;
```
We can just do `return false;`

susiehgt committed:
Adding sorting and grouping to URL

susiehgt committed:
Merge remote-tracking branch 'origin/develop' into develop

susiehgt committed:
Merge branch 'develop' into group-by-orthologs

# Conflicts:
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.css
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.html

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html:
```diff
@@ -84,28 +85,28 @@
               </td>
-              @for (complex of enrichedComplexes; track complex.complex.complexAC) {
+              @for (complex of navigatorComplexes; track complex.complex.complexAC) {
                 <td class="intStoich" [ngClass]="{'intStoichFixedWidth': complexes().length > 6}">
```
You need to update this td element to support predicted subcomponents
```
                <td class="intStoich" [ngClass]="{'intStoichFixedWidth': complexes().length > 6}"
                    [class.predicted]="complex.complex.predictedComplex">
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.ts:
```diff
@@ -1,24 +1,40 @@
 })
 export class ComplexNavigatorButtonsComponent implements OnInit {
-  interactorsSorting = model<string>();
+  componentsSorting = model<NavigatorComponentSorting>();
```
We don't need this anymore, as we have the output. You can delete this, delete `this.componentsSorting.set(typeOfSorting);` on line 37 and delete `[(componentsSorting)]="componentsSorting"` in complex-navigator.component.html (line 4).

jmedinaebi commented:
In the basket, update the complex navigator component.

In basket.component.html (line 52)
```
              <cp-complex-navigator
                [complexSearch]="complexSearchBasket"
                [interactors]="allInteractorsInComplexSearchBasket"
                [canAddComplexesToBasket]="false"
                [canRemoveComplexesFromBasket]="true"
                [(componentsGrouping)]="componentsGrouping"
                [(componentsSorting)]="componentsSorting"
                (onComplexRemovedFromBasket)="deleteComplexFromBasket($event)">
              </cp-complex-navigator>
```

And then in basket.component.ts, add
```
  componentsGrouping = NavigatorComponentGrouping.DEFAULT;
  componentsSorting = NavigatorComponentSorting.DEFAULT;

```

susiehgt committed:
Grouping in basket;
Fixing color of predicted ortholog group components;
Removal of some useless lines;

jmedinaebi commented:
Regarding adding the type of display to the parameters, we should also add a parameter for view vs navigator view, and remove the fragment.

To do that, in `complex-results-component.ts` we do:

1. Update `ngOnInit` to not parse from the fragment, but to get the display mode from the params
   ```
   ngOnInit() {
    this.titleService.setTitle('Complex Portal - Results');
    this.allInteractorsInComplexSearch = [];
    this.route.queryParams.pipe().subscribe(queryParams => {
      this.query = queryParams['query'];
      Object.keys(this.filters).forEach(filter => this.filters[filter] = this.decodeURL(filter, queryParams));
      // Only set the mode if in the params, otherwise rely on the default behaviour
      if (!!queryParams['displayMode']) {
        this._displayType = queryParams['mode'];
      }
      this.confidenceFilter = queryParams['minConfidence'] ? Number(queryParams['minConfidence']) : 1;
      this.componentsGrouping = queryParams['componentsGrouping'] ? queryParams['componentsGrouping'] : NavigatorComponentGrouping.DEFAULT;
      this.componentsSorting = queryParams['componentsSorting'] ? queryParams['componentsSorting'] : NavigatorComponentSorting.DEFAULT;
      this.typeOfDisplay = queryParams['typeOfDisplay'] ? queryParams['typeOfDisplay'] : NavigatorDisplayType.DETAILED;
      this.currentPageIndex = queryParams['page'] ? Number(queryParams['page']) : 1;
      // TODO This is out for now, but CP-84 (JIRA )should fix that!!
      // this.pageSize = queryParams['size'] ? Number(queryParams['size']) : 10;
      this.requestComplexResults();
      document.body.scrollTop = 0;
    });
   }
   ```

2. Update `reloadPage` to set a new param for the type of display
   ```
   queryParams['displayMode'] = this._displayType;
   ```

3. Update `reloadPage` to remove setting the fragment on the navigate call
   ```
   this.router.navigate([], { queryParams });
   ```

jmedinaebi commented:
And to add the display options to the params, do:

1. Create a new enum
   ```
   export enum NavigatorDisplayType {
     COMPACT = 'compact',
     DETAILED = 'detailed'
   }
   ```

2. Set the new type in `complex-results.component.ts`, same as we do for the grouping and sorting. A new variable, setting it from the params in ngOnInit, setting the param in reloadPage and a new method like `onComplexNavigatorGroupingChange` for the new variable.

3. Update `ComplexNavigatorComponent` with a new model for the new variable, same as `componentsSorting` or `componentsGrouping` and a new method like `onDisplayTypeChange`.

4. Update `ComplexNavigatorButtonsComponent` with a new input `typeOfDisplay = input<NavigatorDisplayType>();`

5. Remove `ngOnInit` from `ComplexNavigatorButtonsComponent`

6. Add a new method, like `onGroupingChanged` to `ComplexNavigatorButtonsComponent` for the new variable.

7. Update where `typeOfDisplay` is used in `ComplexNavigatorButtonsComponent` as it is now an input and requires (), and also where it is set, as we should simply call the new on??? method.

8. Replace strings 'compact' and 'detailed' with the enum.

susiehgt committed:
Adding of the display in the URL

susiehgt committed:
Adding of the display in the URL; basket bug fixing

susiehgt committed:
Cleaning and fixing typo

jmedinaebi commented on src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.html:
```diff
@@ -19,14 +30,14 @@
         <label>
           <input type="radio" name="displayType" value="Compact" class="checkbox compact"
-                 (click)="setDisplayType('compact')" [checked]="typeOfDisplay === 'compact'">
+                 (click)="setDisplayType('compact')" [checked]="typeOfDisplay() === NavigatorDisplayType.COMPACT">
```
Replace
```
setDisplayType('compact')
```

With 
```
setDisplayType(NavigatorDisplayType.COMPACT)
```

Same for line 40.

jmedinaebi replied:
We also need to update the type of the argument in `setDisplayType`
```
setDisplayType(type: NavigatorDisplayType) {
```

And update the ifs in that method to use the enum.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -15,8 +24,139 @@ export class ComplexNavigatorComponent {
   onComplexRemovedFromBasket = output<string>();
 
-  interactorsSorting = 'Occurrence';
+  componentsSorting = model<NavigatorComponentSorting>();
+  componentsGrouping = model<NavigatorComponentGrouping>();
+  displayType = model<NavigatorDisplayType>();
```
These 3 can be input instead of model. We already have outputs to emit events, so we don't need to use model here.

jmedinaebi replied:
We would need to update the html (basket and complex-results) where `cp-complex-navigator` element is used to use `[` instead of `[(`.

jmedinaebi replied:
Actually ignore this, leave them as model.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -15,8 +24,139 @@ export class ComplexNavigatorComponent {
   interactorTypeDisplay = true;
   idDisplay = true;
+
+  _allChanges: OutputRef<any>[] = [
```
No longer used, delete.

jmedinaebi replied:
Ignore this, we can simplify all this logic. As mentioned above, keep the models. Then, remove the 3 outputs `onGroupingChange`, `onSortingChange` and `onDisplayTypeChange`; and have only 1 output `anyChange = output<void>();`

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -15,8 +24,139 @@ export class ComplexNavigatorComponent {
+  }
+
+  onGroupingChanged(componentsGrouping: NavigatorComponentGrouping) {
+    this.componentsGrouping.set(componentsGrouping);
```
Remove this set call. Same for the set calls in `onSortingChanged` and `onDisplayTypeChanged`. We send the events and then the other components update this, so we don't need it.

jmedinaebi replied:
Ignore, do not remove the set calls. Keep them, but replace the `this.onGroupingChange.emit(componentsGrouping);` call with `this.anyChange.emit();`. Same on the 2 other methods below.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -47,19 +50,14 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
+      this.componentsGrouping = queryParams['componentsGrouping'] ? queryParams['componentsGrouping'] : NavigatorComponentGrouping.DEFAULT;
+      this.componentsSorting = queryParams['componentsSorting'] ? queryParams['componentsSorting'] : NavigatorComponentSorting.DEFAULT;
+      this.typeOfDisplay = queryParams['typeOfDisplay'] ? queryParams['typeOfDisplay'] : NavigatorDisplayType.DETAILED;
+      this._displayType = queryParams['displayMode'] ? queryParams['displayMode'] : queryParams['mode'];
```
```
      if (!!queryParams['displayMode']) {
        this._displayType = queryParams['displayMode'];
      }
```

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.html:
```diff
@@ -43,19 +43,25 @@ <h2 class="padding-left-large">Total number of results: {{ complexSearch.totalNu
+                                      [(componentsGrouping)]="componentsGrouping"
+                                      [(componentsSorting)]="componentsSorting"
+                                      [(displayType)]="typeOfDisplay"
+                                      (onGroupingChange)=onComplexNavigatorGroupingChange(componentsGrouping)
+                                      (onSortingChange)=onComplexNavigatorSortingChange(componentsSorting)
+                                      (onDisplayTypeChange)="onComplexNavigatorDisplayChange(typeOfDisplay)">
```
Replace these 3 lines with 
```
(anyChange)=reloadPage()>
```

`reloadPage` needs to be made public.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -136,6 +137,21 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
     this.reloadPage();
   }
 
+  public onComplexNavigatorGroupingChange(grouping: NavigatorComponentGrouping) {
```
Delete these 3 methods. No longer needed after the other changes mentioned on other comments.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-ortholog-group.model.ts:
```diff
@@ -0,0 +1,127 @@
+  private _subComponents: INavigatorComponent[];
+  private _timesAppearing: number;
+  private _indexAppearing: number;
+  private _isOrthologGroup: boolean;
```
We already return always true, we don't need this variable.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-simple-component.model.ts:
```diff
@@ -0,0 +1,118 @@
+  private _subComponents: ComplexComponent[];
+  private _timesAppearing: number;
+  private _indexAppearing: number;
+  private _isOrthologGroup: boolean;
```
We already return always false, we don't need this variable.

susiehgt committed:
Removing useless functions and variables;
Using the correct variable in the HTML;

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator.component.ts:
```diff
@@ -13,10 +22,132 @@ export class ComplexNavigatorComponent {
+  componentsGrouping = model<NavigatorComponentGrouping>();
+  displayType = model<NavigatorDisplayType>();
   onComplexRemovedFromBasket = output<string>();
+  onGroupingChange = output<NavigatorComponentGrouping>();
+  onSortingChange = output<NavigatorComponentSorting>();
+  onDisplayTypeChange = output<NavigatorDisplayType>();
```
These 3 outputs can be deleted, no longer used.

jmedinaebi reviewed (commented):
I managed to get the search for panther xrefs working. It is not one of the default fields we search for, but it works if we specify the field. Meaning, if we search for `PTHR11973` we get nothing, but if we search for `pxref:("PTHR11973")` we do.

So, to make this work nicely.

1. Update `INavigatorSubComponent` with a new get method.
   ```
   get componentQuery(): string;
   ```

2. Update `NavigatorSimpleComponent` to just return the id on that method.
   ```
   get componentQuery(): string {
     return this.identifier;
   }

   ```

3. Update `NavigatorOrthologGroup` to return the proper query we need
   ```
   get componentQuery(): string {
     return 'pxref:("' + this.identifier + '")';
   }
   ```

4. Update `table-interactor-name.component` component with a new input
   ```
   interactorQuery = input<string>();
   ```

5. Update the html of that component to use the input on the search (lines 41 and 60)
   ```
   [queryParams]="{query: interactorQuery(), page: 1}"
   ```

6. Update `table-interactor-column.component` component to pass the new field to the component updated before
   6.1. For the main interactors (line 26)
     ```
     ...
              [interactorId]="interactor.identifier"
              [interactorQuery]="interactor.componentQuery"
     ...
     ```
   6.2. For the subcomponents we can use the identifier as we currently do
     ```
     ...
                    [interactorId]="subComponent.identifier"
                    [interactorQuery]="subComponent.identifier"
     ...
     ```

susiehgt committed:
Adding xref search;
Cleaning;

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-simple-component.model.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  get orthologsGroup(): XRef {
+    for (const xref of this._interactor.xrefs) {
```
In case we want to deploy these changes before the back-end changes, which we might, we need to update this.

The back-end (production one) is not returning xrefs now, so `interactor.xrefs` would be undefined. So we need to do:

```
    if (!!this._interactor.xrefs) {
      for (const xref of this._interactor.xrefs) {
        if (!!xref && xref.database === 'panther') {
          return xref;
        }
      }
    }
    return null;
```

Basically adding the if at the beginning to be safe.

jmedinaebi reviewed (commented):
Really good, just one small thing before we're ready to merge this.

susiehgt committed:
Small fixes

susiehgt committed:
Refactoring Susie

susiehgt committed:
Changes by eragueneau

jmedinaebi commented on src/app/app.component.ts:
```diff
@@ -85,7 +85,7 @@ export class AppComponent implements OnInit, AfterViewInit {
   //     }
   //     if (localMastheadImage != null) {
-  //       localMasthead.style.backgroundImage = 'url(' + localMastheadImage.getAttribute("content") + ')';
+  //       localMasthead.style.backgroundImage = 'state(' + localMastheadImage.getAttribute("content") + ')';
```
I'm guessing you did a replace all and this got replace by accident. It is not relevant as it is commented out, but revert it anyway.

jmedinaebi commented on src/app/basket/basket.component.css:
```diff
@@ -1,9 +1,3 @@
-.ComplexNavigator, .hover {
```
Why was this removed?

susiehgt replied:
Eliot used the medium-12 in the HTML so I think he removed it because it is not used anymore

EliotRagueneau replied:
Yes basically before, we were limiting the size of the lements themeselves, but I neeeded to restrict the size of the prent really, so that the tabs match the content. Then i could have just moved that up, but basically, 70% can be problematic when on small screens, so I used indeed teh small-12 medium-10 large-8 I think to ensure that in all scales, we have enough space

jmedinaebi commented on src/app/complex/complex-results/complex-filter/complex-filter.component.ts:
```diff
@@ -43,7 +43,10 @@ export class ComplexFilterComponent implements OnInit {
   onResetAllFilters = output<boolean>();
 
-  confidenceGap = input<number>(2);
+  /**
```
Can you add a bit more? Minimum 5, otherwise it breaks? What's 5?

susiehgt replied:
@EliotRagueneau

EliotRagueneau replied:
Yep, basically the slider has a minimum length that I couldn't overwrite, at least easily, so the minimum space between stars in order for the star pattern to take as much space as the slider is 5. If you go bellow, the area gets ddimisnished, but not the slider, leading it to be over the next area, and not alligend with the stars.
5 is pixel size

jmedinaebi replied:
Then the comment should be something like
```
Slider length cannot be overwritten. In order for the stars to take the same space as the slider, the minimum space between stars must be 5, otherwise they don't align.
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-component.model.ts:
```diff
@@ -0,0 +1,51 @@
+
+  set indexAppearing(value: number);
+
+  componentIds(): string[];
```
To be consistent, make it `get componentIds(): string[];`. Then you need to replace where it is called, from `componentIds()` to `componentIds`. And also change the classes that implement this interface.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/service/state/complex-navigator-display.service.ts:
```diff
@@ -0,0 +1,66 @@
+    this.route.queryParams.subscribe(queryParams => {
+      for (const param of Object.keys(this.params())) {
+        if (!queryParams[param]) {
+          continue;
```
We can read from local storage here. If there's no query params for a param, then we fetch from local storage.

jmedinaebi replied:
The other part is to save to the local storage. That I'm not sure how at the moment.

EliotRagueneau replied:
We can just add it to the other effect we have to write to the localStorage

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -32,34 +31,26 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
-  private _navigatorLastPageIndex;
-
-  constructor(private route: ActivatedRoute, private router: Router,
+  private pageSize = 20; // This is where we set the size of the pages for navigator view
```
Comment is no longer true. We're now using page size 20 for both views.

EliotRagueneau replied:
Yes we actually removed that because it was a lot of complexity and reloading time just to have 5 iteems less in the list view, which is not that important in my opinion

jmedinaebi replied:
That makes sense, but the comment is not updated.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/model/navigator-simple-component.model.ts:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  get orthologsGroup(): XRef {
+    return this._interactor.xrefs?.find(xref => xref.qualifier === 'orthology-group');
```
Just qualifier? No check for the database?

EliotRagueneau replied:
I prefer to use the qualifier if we decide to change the source orthology database at some point, since they are many alternatives, and there is no "clear winner" of who's the best

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html:
```diff
@@ -46,66 +38,60 @@
               }
-              @if (enrichedComplexes.length > 6) {
+              @if (navigatorComplexes.length > 6) {
                 <div class="spaceHolder"></div>
```
```
<div [class.sorting]="isSorting()" class="spaceHolder"></div>
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html:
```diff
@@ -22,22 +22,14 @@
           <td class="interactorsColumn"
```
```
<td class="interactorsColumn horizontalHeader"
              [class.fixedWidth]="fixedWidth()"
              [style]="ranges.length!=0 ? 'left:3ch':''">
```

jmedinaebi replied:
The CSS file also needs to be updated.

Currently, it is like this, which it is not working.
```
.horizontalHeader {
  min-width: 114px;
  max-width: 114px;
}

.horizontalHeader.fixedWidth {
  width: 230px;
}
```

It works if you do:
```
.horizontalHeader {
  min-width: 114px;
  max-width: 114px;
}

.horizontalHeader.fixedWidth {
  width: 230px;
  min-width: 230px;
  max-width: 230px;
}
```

jmedinaebi reviewed (commented):
Looking good. I left a few comments on some small things.

EliotRagueneau committed:
:memo: Improve doc on confidence

EliotRagueneau committed:
:bulb: remove pageSize comment which is no longer true

susiehgt committed:
Adding display option in the local storage;
Adding of google analytics to buttons;

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/service/state/complex-navigator-display.service.ts:
```diff
@@ -0,0 +1,95 @@
+      for (const param of paramKeys) {
+        let value: any;
+
+        // 1. Check URL params first
+        if (queryParams[param] !== undefined) {
+          switch (typeof this[param]()) {
+            case 'boolean':
+              value = queryParams[param] === 'true';
+              break;
+            default:
+              value = queryParams[param];
+          }
+          // 2. Then heck local storage
+        } else if (localStorage.getItem(param) !== null) {
+          switch (typeof this[param]()) {
+            case 'boolean':
+              value = localStorage.getItem(param) === 'true';
+              break;
+            default:
+              value = localStorage.getItem(param);
+          }
```
We can create a private method parseParam
```
private parseParam(paramName, paramValue): any {
  switch (typeof this[paramName]()) {
    case 'boolean':
      return paramValue === 'true';
    default:
      return paramValue;
  }
}
```

And then, simplify this code:
```
        // 1. Check URL params first
        if (queryParams[param] !== undefined) {
          value = this.parseParam(param, queryParams[param]);
          // 2. Then heck local storage
        } else if (localStorage.getItem(param) !== null) {
          value = this.parseParam(param, localStorage.getItem(param));
        } else {
          ...
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/service/state/complex-navigator-display.service.ts:
```diff
@@ -0,0 +1,95 @@
+    });
+
+    effect(() => {
+      if (this.ignore) {
```
What's the ignore variable?

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/service/state/complex-navigator-display.service.ts:
```diff
@@ -0,0 +1,95 @@
+
+
+    effect(() => {
+      this.updateLocalStorage();
```
Instead of having 2 effects, we could just have one, right?

```
    effect(() => {
      if (this.ignore) {
        return;
      }
      this.updateLocalStorage();
      this.router.navigate([], {queryParams: this.params(), queryParamsHandling: 'merge'});
    }, {allowSignalWrites: true});
```

jmedinaebi commented on src/app/shared/google-analytics/service/analytics.service.ts:
```diff
@@ -70,4 +70,12 @@ export class AnalyticsService {
     this.analytics.event(Action[Action.RatingComplexNavigator], Category[Category.complexNavigator], 'rating', rating);
   }
+
+  public useOrthologyButton(action: Action) {
```
Rename this method, as it is more generic than ortholog button. It can be `useComplexNavigatorSettingButton`.

jmedinaebi replied:
Rename this method, as it is more generic than ortholog button. It can be useComplexNavigatorSettingButton.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/service/state/complex-navigator-display.service.ts:
```diff
@@ -0,0 +1,95 @@
+            default:
+              value = queryParams[param];
+          }
+          // 2. Then heck local storage
```
Typo, I guess you meant `Then check...`

EliotRagueneau committed:
Merge remote-tracking branch 'origin/group-by-orthologs' into group-by-orthologs

# Conflicts:
#	src/app/complex/complex-results/complex-results.component.ts

EliotRagueneau committed:
:bug: Restore sticky header

EliotRagueneau committed:
:truck: Rename merge orthologs to group by orthologs + ensure menu position

EliotRagueneau committed:
:lipstick: Fix column width

susiehgt committed:
Adding the shadows in complex nav

EliotRagueneau committed:
:sparkles: Control shadow appearance based on scroll

EliotRagueneau committed:
:bug: fix something

susiehgt committed:
Top shadow complex nav

EliotRagueneau committed:
Merge remote-tracking branch 'origin/group-by-orthologs' into group-by-orthologs

# Conflicts:
#	src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.css
#	src/styles.scss

EliotRagueneau committed:
:bug: fix make sticky synced header work with shadows

EliotRagueneau committed:
:bug: fix missing stoichiometry because backend doesn't know how to spell ^^

EliotRagueneau committed:
:lipstick: Remove border bottom as shadow is enough

EliotRagueneau committed:
:art: Format

EliotRagueneau committed:
:sparkles: Init correctly shadows initially

EliotRagueneau committed:
:lipstick: Avoid permanent vertical scrollbar on scroller

jmedinaebi commented:
Apart from the comments above, a few other things regarding the table style.

1. Something changed about the width that broke the table. I think it happens with and without orthologs, and with or without sorting.
![Screenshot 2024-10-01 at 10 53 59](https://github.com/user-attachments/assets/b2905004-8570-4287-be14-6aeefabb6a1e)

2. The shadow of the header extends all the way to the right
![Screenshot 2024-10-01 at 10 54 06](https://github.com/user-attachments/assets/30144017-abb7-4665-b40a-918c330d9595)

3. The shadow is not displayed when we're at the very top of the table, so the separation between the header and the first row is not easy to see
![Screenshot 2024-10-01 at 10 54 17](https://github.com/user-attachments/assets/72a4503d-ce3d-4d91-b1e4-0a677a8f9652)

I'll take a look to see if I can find out how to fix these issues, so you can focus on the xml maker.

jmedinaebi committed:
Rename useOrthologyButton method

jmedinaebi committed:
Make componentIds a get method

jmedinaebi committed:
Display humap xref on header

jmedinaebi committed:
Reduce duplicated code

jmedinaebi committed:
Fix style issues on table width and shadow

EliotRagueneau commented:
![Screenshot 2024-10-01 at 12 58 26](https://github.com/user-attachments/assets/1cf25344-3924-4ac5-bb93-997d043fe574)
I think the fix you did on shadows made things even worse

jmedinaebi commented:
> ![Screenshot 2024-10-01 at 12 58 26](https://private-user-images.githubusercontent.com/42945601/372457782-1cf25344-3924-4ac5-bb93-997d043fe574.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3ODQzMDMsIm5iZiI6MTcyNzc4NDAwMywicGF0aCI6Ii80Mjk0NTYwMS8zNzI0NTc3ODItMWNmMjUzNDQtMzkyNC00YWM1LWJiOTMtOTk3ZDA0M2ZlNTc0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDEyMDAwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkMDhhMjk3ZGUzODBjNTg1YmY3OTIzOGZlZGQ0MDdjZTA4ODkwOWE4YzliNzExNWUxNGNkMmIyOTZiNWZjMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xY_rTyhXkqZpYWWdokYCp3hLXSaqy6jE6qqfepIvi_M) I think the fix you did on shadows made things even worse

I see what's happening. I think I know why it's happening, so I'll take a look.

EliotRagueneau commented:
Normally you don't have to precise the shadow's width, it's just folllowing its container size, so all you neeed to do is have the parent the right size really

jmedinaebi commented:
> Normally you don't have to precise the shadow's width, it's just folllowing its container size, so all you neeed to do is have the parent the right size really

That's what I was thinking on doing.

EliotRagueneau committed:
:lipstick: Avoid nav overflowing page with vw

EliotRagueneau committed:
:lipstick: Go back to parent size based width for horizontal shadow

EliotRagueneau committed:
:lipstick: Perfect spaceholder width

jmedinaebi committed:
Set navigator table width on header

EliotRagueneau committed:
:recycle: Use scss + use css variable inheritance instead of recalculating + fix little gap to the right of header

EliotRagueneau committed:
Merge remote-tracking branch 'origin/group-by-orthologs' into group-by-orthologs

# Conflicts:
#	src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.css
#	src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.html
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.scss

EliotRagueneau committed:
:bug: remove no longer required sorting class

jmedinaebi committed:
Fix issues with list and navigator view

jmedinaebi committed:
Small fixed on table width

EliotRagueneau commented:
@jmedinaebi thanks for the fix on groups, completely missed that! 
I did change quite a lot of the calculations in the css to only do it in the parrent and use the result everywhere. I'm not sure it's actually optimising since it looks like the variable contains the calc and not the result, but at least I find it clearer and avoid code duplication :D
I think we can aceept this PR now, would you agree?

jmedinaebi commented:
> @jmedinaebi thanks for the fix on groups, completely missed that! I did change quite a lot of the calculations in the css to only do it in the parrent and use the result everywhere. I'm not sure it's actually optimising since it looks like the variable contains the calc and not the result, but at least I find it clearer and avoid code duplication :D I think we can aceept this PR now, would you agree?

I think it is ready.

@EliotRagueneau @susiehgt 
Complex indexer has finished and I have updated SOLR in the public instances too. Public Complex web services are now returning orthology groups, and I have deployed these changes on the github pages with the production environment instead of the test one. You can now check these changes with production data, like this search https://complex-portal.github.io/complex-portal-view/complex/search?query=*&page=1&minConfidence=1&displayMode=navigator-view&mergeOrthologs=true&componentsSorting=Default&idDisplay=true&organismIconDisplay=true&typeIconDisplay=true&predicted=true
